### ISSUE TITLE
Weird behaviour of field_simplify with 1/1

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#9370]
Dear Coq Team,

the following sequence of Coq commands results in the goals in the comments:
```
Require Import Reals.
Open Scope R_scope.
Goal 1/1=1.
Proof.
  field_simplify(1/1). (* 1 / 1 = 1 *)
  field_simplify. (* 1 / 1 / (1 / 1) = 1 / 1 *)
  field_simplify. (* 1 / 1 = 1 / 1 *)
```

I would expect that field_simplify (1/1) simplifies 1/1 to 1. The following field_simplify is rather weird. Of cause field can solve this, but I frequently use field_simplify for subexpressions of terms field can't handle.

Otherwise I am quite happy with field_simplify, although I think one might be able to extend it to what radcan does in Maxima (which also takes care of roots and powers and usually gets the job done).

This is on Coq 8.9 master from a week back,

Best regards,

Michael

**Comment by gares** [comment]
CC @thery

**Comment by thery** [comment]
@MSoegtropIMC 
From what I remember, field_simplify always returns a fraction. We simply didn't trigger the trivial simplification (a/1 = a), maybe we should

If you do field_simplify (2/2) it will return 1/1

Your second example looks like a bug to me, I will look into it.

Your third example has the same explanation then (1/1).

Do you sometime use field_simplify_eq?

**Comment by MSoegtropIMC** [comment]
@thery : thanks for the hint to field_simplify_eq - I am sure I can make good use of it. But in this case I had to simplify terms deep inside nested trigonometric and root functions, so that field_simplify is the only tool that gets the job done.

> We simply didn't trigger the trivial simplification (a/1 = a), maybe we should

Yes please. It is of cause not a big deal, but it is not nice and doesn't leave a good first impression of this very nice and useful tactic - it simply looks rather silly to the human eye.

Also in practice it would help. I had e.g. the case that I want to simplify sqrt(1/1) to 1 and wanted to use field_simplify to get it to sqrt(1) so that I can rewrite with the lemma that this equals 1. But I first had to manually rewrite with a x/1=x lemma, which is a bit of a nuisance.

The 2nd case needlessly further reduces the users trust that this tactic is the right tool to get field related simplifications done effectively.

As I said, this is all just first glance stuff - otherwise the tactic works very well, but for humans the first impression is sometimes more important than the depth behind something.

**Comment by thery** [comment]
@MSoegtropIMC
As the field tactic let you define a post process, you could remap it to perform your simplification.

```
Lemma Rdiv_1_r r : r / 1 = r.
Proof.
now unfold Rdiv; rewrite Rinv_1, Rmult_1_r.
Qed.

Add Field RField : Rfield
   (completeness Zeq_bool_complete, constants [IZR_tac], power_tac R_power_theory [Rpow_tac],
    postprocess[repeat rewrite Rdiv_1_r]).
```

If you can experiment with that (and maybe even extend), we could integrate it in the standard release.

**Comment by MSoegtropIMC** [comment]
@thery : there is some more weird behavior of field_simplify:

```
Require Import Reals.
Open Scope R_scope.

Goal forall x:R, x>0 -> x = 1 / (1 / sqrt x²).
Proof.
  intros.
  field_simplify.    (* x / 1 = 1 / (1 / sqrt (x / 1)²) *)
  field_simplify_eq. (* x = sqrt (x / 1)² *)
```

field_simplify_eq does a good job, but why does field_simplify not resolve the double fraction? I would interpret the documentation that field_simplify reduces a term (if possible) to a ratio of polynomials such that this should work.

Another case:
```
Goal forall x:R, x>0 -> x = 1 * sqrt (x)² / 1.
Proof.
  intros.
  field_simplify.    (* x / 1 = 1 * sqrt (x / 1)² / 1 *)
```
Why is the 1 * not removed here? 

All this is easy to fix with some pre and post processing, but I wonder if this is the right way to go.

Best regards,

Michael

**Comment by thery** [comment]
@MSoegtropIMC
I think it is because you got two rewriting x -> x/1 and (1 / 1 (... x)) -> z that are not independent so
the application of the first one forbids the application of the second one. Sad story.
Try this:
> 
> Goal forall x:R, x>0 -> x = 1 / (1 / sqrt x²).
> Proof.
>   intros.
>   apply sym_equal.
>   field_simplify.
>

**Comment by andres-erbsen** [comment]
@MSoegtropIMC you might be interested in the following technique I use to run simplification tactics inside expressions:

```coq
Require Import Reals.
Open Scope R_scope.

Ltac set_evars := repeat match goal with |- context [?e] => is_evar e; set e end.
Ltac subst_evars := repeat match goal with x := ?e |- _ => is_evar e; subst e end.
Ltac sane_field_simplify_eq := set_evars; field_simplify_eq; subst_evars.

Tactic Notation "simplify" open_constr(pat) "by" tactic3(tac) :=
  unshelve erewrite (_: pat = _); shelve_unifiable; [tac; exact eq_refl|].
  
Goal forall x:R, x>0 -> x = 1 * sqrt (x)² / 1.
Proof.
  intros.
  simplify (_/1) by sane_field_simplify_eq.
```

**Comment by thery** [comment]
@andres-erbsen 

nice but it still seems a bit fragile
```  
Goal forall x:R, x>0 -> 1 = (1 * sqrt (x)²) / (x + 1 - 1).
Proof.
  intros.
  simplify (_ / _) by sane_field_simplify_eq.
```

**Comment by maximedenes** [comment]
@thery and I have a tentative fix in #9854.
